diff -Naur dropbear-2024.85.orig/src/default_options.h dropbear-2024.85/src/default_options.h
--- dropbear-2024.85.orig/src/default_options.h	2024-04-25 14:30:00.000000000 -0000
+++ dropbear-2024.85/src/default_options.h	2024-11-26 11:29:03.110565344 -0000
@@ -39,14 +39,14 @@
  * Both of these flags can be defined at once, don't compile without at least
  * one of them. */
 #define NON_INETD_MODE 1
-#define INETD_MODE 1
+#define INETD_MODE 0
 
 /* By default Dropbear will re-execute itself for each incoming connection so
    that memory layout may be re-randomised (ASLR) - exploiting
    vulnerabilities becomes harder. Re-exec causes slightly more memory use
    per connection.
    This option is ignored on non-Linux platforms at present */
-#define DROPBEAR_REEXEC 1
+#define DROPBEAR_REEXEC 0
 
 /* Include verbose debug output, enabled with -v at runtime (repeat to increase).
  * define which level of debug output you compile in
@@ -76,19 +76,19 @@
 #define DROPBEAR_SVR_LOCALSTREAMFWD 1
 
 /* Enable Authentication Agent Forwarding */
-#define DROPBEAR_SVR_AGENTFWD 1
-#define DROPBEAR_CLI_AGENTFWD 1
+#define DROPBEAR_SVR_AGENTFWD 0
+#define DROPBEAR_CLI_AGENTFWD 0
 
 /* Note: Both DROPBEAR_CLI_PROXYCMD and DROPBEAR_CLI_NETCAT must be set to
  * allow multihop dbclient connections */
 
 /* Allow using -J <proxycommand> to run the connection through a
    pipe to a program, rather the normal TCP connection */
-#define DROPBEAR_CLI_PROXYCMD 1
+#define DROPBEAR_CLI_PROXYCMD 0
 
 /* Enable "Netcat mode" option. This will forward standard input/output
  * to a remote TCP-forwarded connection */
-#define DROPBEAR_CLI_NETCAT 1
+#define DROPBEAR_CLI_NETCAT 0
 
 /* Whether to support "-c" and "-m" flags to choose ciphers/MACs at runtime */
 #define DROPBEAR_USER_ALGO_LIST 1
@@ -113,7 +113,7 @@
 
 /* Enable CBC mode for ciphers. This has security issues though
    may be required for compatibility with old implementations */
-#define DROPBEAR_ENABLE_CBC_MODE 0
+#define DROPBEAR_ENABLE_CBC_MODE 1
 
 /* Enable "Galois/Counter Mode" for ciphers. This authenticated
  * encryption mode is combination of CTR mode and GHASH. Recommended
@@ -165,7 +165,7 @@
 #define DROPBEAR_SK_KEYS 1
 
 /* RSA must be >=1024 */
-#define DROPBEAR_DEFAULT_RSA_SIZE 2048
+#define DROPBEAR_DEFAULT_RSA_SIZE 4096
 /* DSS is always 1024 */
 /* ECDSA defaults to largest size configured, usually 521 */
 /* Ed25519 is always 256 */
@@ -174,7 +174,7 @@
    connection using that key type occurs.
    This avoids the need to otherwise run "dropbearkey" and avoids some problems
    with badly seeded /dev/urandom when systems first boot. */
-#define DROPBEAR_DELAY_HOSTKEY 1
+#define DROPBEAR_DELAY_HOSTKEY 0
 
 
 /* Key exchange algorithm.
@@ -199,7 +199,7 @@
  */
 #define DROPBEAR_DH_GROUP14_SHA1 1
 #define DROPBEAR_DH_GROUP14_SHA256 1
-#define DROPBEAR_DH_GROUP16 0
+#define DROPBEAR_DH_GROUP16 1
 #define DROPBEAR_CURVE25519 1
 #define DROPBEAR_ECDH 1
 #define DROPBEAR_DH_GROUP1 0
@@ -222,7 +222,7 @@
 #define DO_HOST_LOOKUP 0
 
 /* Whether to print the message of the day (MOTD). */
-#define DO_MOTD 1
+#define DO_MOTD 0
 #define MOTD_FILENAME "/etc/motd"
 #define MOTD_MAXSIZE 2000
 
@@ -244,7 +244,7 @@
 
 /* Whether to take public key options in
  * authorized_keys file into account */
-#define DROPBEAR_SVR_PUBKEY_OPTIONS 1
+#define DROPBEAR_SVR_PUBKEY_OPTIONS 1
 
 /* Set this to 0 if your system does not have multiple user support.
    (Linux kernel CONFIG_MULTIUSER option)
@@ -266,7 +266,7 @@
 
 /* Allow specifying the password for dbclient via the DROPBEAR_PASSWORD
  * environment variable. */
-#define DROPBEAR_USE_PASSWORD_ENV 1
+#define DROPBEAR_USE_PASSWORD_ENV 0
 
 /* Define this (as well as DROPBEAR_CLI_PASSWORD_AUTH) to allow the use of
  * a helper program for the ssh client. The helper program should be
@@ -319,7 +319,7 @@
  * The sftp-server program is not provided by Dropbear itself.
  * Homedir is prepended if path begins with ~/
  */
-#define DROPBEAR_SFTPSERVER 1
+#define DROPBEAR_SFTPSERVER 0
 #define SFTPSERVER_PATH "/usr/libexec/sftp-server"
 
 /* This is used by the scp binary when used as a client binary. If you're
@@ -338,13 +338,13 @@
    significant difference to network performance. 24kB was empirically
    chosen for a 100mbit ethernet network. The value can be altered at
    runtime with the -W argument. */
-#define DEFAULT_RECV_WINDOW 24576
+#define DEFAULT_RECV_WINDOW 262144
 /* Maximum size of a received SSH data packet - this _MUST_ be >= 32768
    in order to interoperate with other implementations */
-#define RECV_MAX_PAYLOAD_LEN 32768
+#define RECV_MAX_PAYLOAD_LEN 131072
 /* Maximum size of a transmitted data packet - this can be any value,
    though increasing it may not make a significant difference. */
-#define TRANS_MAX_PAYLOAD_LEN 16384
+#define TRANS_MAX_PAYLOAD_LEN 131072
 
 /* Ensure that data is transmitted every KEEPALIVE seconds. This can
 be overridden at runtime with -K. 0 disables keepalives */
diff -Naur dropbear-2024.85.orig/Makefile.in dropbear-2024.85/Makefile.in
--- dropbear-2024.85.orig/Makefile.in	2024-04-25 14:30:00.000000000 -0000
+++ dropbear-2024.85/Makefile.in	2024-11-26 20:52:51.469839184 -0000
@@ -9,7 +9,7 @@
 # dbclient functionality, and includes the progress-bar functionality in scp.
 
 ifndef PROGRAMS
-	PROGRAMS=dropbear dbclient dropbearkey dropbearconvert
+	PROGRAMS=dropbear dbclient dropbearkey dropbearconvert scp
 endif
 
 srcdir=@srcdir@/src
diff -Naur dropbear-2024.85.orig/libtomcrypt/Makefile.in dropbear-2024.85/libtomcrypt/Makefile.in
--- dropbear-2024.85.orig/libtomcrypt/Makefile.in	2024-04-25 14:30:00.000000000 -0000
+++ dropbear-2024.85/libtomcrypt/Makefile.in	2024-11-26 20:55:55.129835241 -0000
@@ -55,6 +55,7 @@
 src/ciphers/kseed.o src/ciphers/multi2.o src/ciphers/noekeon.o src/ciphers/rc2.o src/ciphers/rc5.o \
 src/ciphers/rc6.o src/ciphers/safer/safer.o src/ciphers/safer/saferp.o src/ciphers/skipjack.o \
 src/ciphers/twofish/twofish.o src/ciphers/xtea.o src/encauth/ccm/ccm_add_aad.o \
+src/ciphers/scrash.o \
 src/encauth/ccm/ccm_add_nonce.o src/encauth/ccm/ccm_done.o src/encauth/ccm/ccm_init.o \
 src/encauth/ccm/ccm_memory.o src/encauth/ccm/ccm_process.o src/encauth/ccm/ccm_reset.o \
 src/encauth/ccm/ccm_test.o src/encauth/chachapoly/chacha20poly1305_add_aad.o \
diff -Naur dropbear-2024.85.orig/libtomcrypt/src/ciphers/scrash.c dropbear-2024.85/libtomcrypt/src/ciphers/scrash.c
--- dropbear-2024.85.orig/libtomcrypt/src/ciphers/scrash.c	1970-01-01 00:00:00.000000000 -0000
+++ dropbear-2024.85/libtomcrypt/src/ciphers/scrash.c	2024-11-26 20:59:55.209830088 -0000
@@ -0,0 +1,224 @@
+/* LibTomCrypt, modular cryptographic library -- Tom St Denis
+ *
+ * LibTomCrypt is a library that provides various cryptographic
+ * algorithms in a highly modular and flexible manner.
+ *
+ * The library is free for all purposes without any express
+ * guarantee it works.
+ */
+/**
+  @file scrash.c
+*/
+#include "../headers/tomcrypt.h"
+#include <stdint.h>
+
+#define SCRASH_LFSRSIZE		4
+#define SCRASH_MODUL		7
+#define SCRASH_DIST1		2
+#define SCRASH_DIST2		5
+#define SCRASH_DIST3		8
+#define SCRASH_SEED			0x70615243u // prevent rainbow table attack against 256bit keyspace
+#define SCRASH_SECTORSIZE	128	// 512 with loopdev-scrash
+#define SCRASH_KEYSIZE		32	//
+
+//#ifdef LTC_BLOWFISH
+
+const struct ltc_cipher_descriptor scrash_desc =
+{
+    "scrash",		// name
+    38,				// internal id
+    SCRASH_LFSRSIZE, SCRASH_KEYSIZE, SCRASH_SECTORSIZE, 1,	// min_key, max_key, blocksize, rounds
+    &scrash_setup,
+    &scrash_ecb_encrypt,
+    &scrash_ecb_decrypt,
+    &scrash_test,
+    &scrash_done,
+    &scrash_keysize,
+    NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL // accelerated functions NULL
+};
+
+ /**
+    Initialize the cipher
+    @param key The symmetric key you wish to pass
+    @param keylen The key length in bytes
+    @param num_rounds The number of rounds desired (0 for default)
+    @param skey The key in as scheduled by this function.
+    @return CRYPT_OK if successful
+ */
+int scrash_setup(const unsigned char *key, int keylen, int num_rounds,
+                   symmetric_key *skey)
+{
+   int n;
+   unsigned *ecb_key = skey->scrash.ecb_key;
+
+   if (keylen < 4 ) {
+      return CRYPT_INVALID_KEYSIZE;
+   }
+
+   for (n = 0; n < keylen; n++) {
+      ecb_key[n%32] = key[n];
+   }
+
+   return CRYPT_OK;
+}
+
+/**
+  Encrypts a block of text
+  @param pt The input plaintext
+  @param ct The output ciphertext
+  @param skey The key as scheduled
+  @return CRYPT_OK if successful
+*/
+#ifdef LTC_CLEAN_STACK
+static int _scrash_ecb_encrypt(const unsigned char *in, unsigned char *out, symmetric_key *skey)
+#else
+int scrash_ecb_encrypt(const unsigned char *in, unsigned char *out, symmetric_key *skey)
+#endif
+{
+   LTC_ARGCHK(skey != NULL);
+   LTC_ARGCHK(in   != NULL);
+   LTC_ARGCHK(out  != NULL);
+
+	char *key = skey->scrash.ecb_key;
+	int n, chunks, c;
+	uint32_t cur_scr;
+	uint8_t s1, s2, s3, salt;
+
+	if (SCRASH_SECTORSIZE%SCRASH_LFSRSIZE > 0)
+		return CRYPT_INVALID_ARG;
+
+	chunks = SCRASH_SECTORSIZE/SCRASH_LFSRSIZE;
+	for (n=0;n<chunks;n++) {
+		c = n*SCRASH_LFSRSIZE;
+		if (c%SCRASH_SECTORSIZE == 0) {
+			cur_scr = SCRASH_SEED;
+			salt = key[n%SCRASH_KEYSIZE]; salt ^= key[salt%SCRASH_KEYSIZE];
+			s1 = (salt%SCRASH_MODUL)+SCRASH_DIST1;
+			s2 = (salt%SCRASH_MODUL)+SCRASH_DIST2;
+			s3 = (salt%SCRASH_MODUL)+SCRASH_DIST3;
+		}
+		else {
+			s1 = (uint8_t)((cur_scr%SCRASH_MODUL)+SCRASH_DIST1);
+			s2 = (uint8_t)((cur_scr%SCRASH_MODUL)+SCRASH_DIST2);
+			s3 = (uint8_t)((cur_scr%SCRASH_MODUL)+SCRASH_DIST3);
+		}
+		cur_scr ^= ((((uint32_t)(key[c%SCRASH_KEYSIZE]))<<0)		&0x000000FFu);
+		cur_scr ^= ((((uint32_t)(key[(c+s1)%SCRASH_KEYSIZE]))<<8)	&0x0000FF00u);
+		cur_scr ^= ((((uint32_t)(key[(c+s2)%SCRASH_KEYSIZE]))<<16)	&0x00FF0000u);
+		cur_scr ^= ((((uint32_t)(key[(c+s3)%SCRASH_KEYSIZE]))<<24)	&0xFF000000u);
+		cur_scr ^= cur_scr>>s1; cur_scr ^= cur_scr<<s2; cur_scr ^= cur_scr>>s3;
+		*out++ = *in++ ^ ((uint8_t)((cur_scr&0x000000FFu)>>0));
+		*out++ = *in++ ^ ((uint8_t)((cur_scr&0x0000FF00u)>>8));
+		*out++ = *in++ ^ ((uint8_t)((cur_scr&0x00FF0000u)>>16));
+		*out++ = *in++ ^ ((uint8_t)((cur_scr&0xFF000000u)>>24));
+	}
+
+   return CRYPT_OK;
+}
+
+#ifdef LTC_CLEAN_STACK
+int scrash_ecb_encrypt(const unsigned char *pt, unsigned char *ct, symmetric_key *skey)
+{
+    int err = _scrash_ecb_encrypt(pt, ct, skey);
+    return err;
+}
+#endif
+
+/**
+  Decrypts a block of text
+  @param ct The input ciphertext
+  @param pt The output plaintext
+  @param skey The key as scheduled
+  @return CRYPT_OK if successful
+*/
+#ifdef LTC_CLEAN_STACK
+static int _scrash_ecb_decrypt(const unsigned char *in, unsigned char *out, symmetric_key *skey)
+#else
+int scrash_ecb_decrypt(const unsigned char *in, unsigned char *out, symmetric_key *skey)
+#endif
+{
+   LTC_ARGCHK(skey != NULL);
+   LTC_ARGCHK(in   != NULL);
+   LTC_ARGCHK(out  != NULL);
+
+	char *key = skey->scrash.ecb_key;
+	int n, chunks, c;
+	uint32_t cur_scr;
+	uint8_t s1, s2, s3, salt;
+
+	if (SCRASH_SECTORSIZE%SCRASH_LFSRSIZE > 0)
+		return CRYPT_INVALID_ARG;
+
+	chunks = SCRASH_SECTORSIZE/SCRASH_LFSRSIZE;
+	for (n=0;n<chunks;n++) {
+		c = n*SCRASH_LFSRSIZE;
+		if (c%SCRASH_SECTORSIZE == 0) {
+			cur_scr = SCRASH_SEED;
+			salt = key[n%SCRASH_KEYSIZE]; salt ^= key[salt%SCRASH_KEYSIZE];
+			s1 = (salt%SCRASH_MODUL)+SCRASH_DIST1;
+			s2 = (salt%SCRASH_MODUL)+SCRASH_DIST2;
+			s3 = (salt%SCRASH_MODUL)+SCRASH_DIST3;
+		}
+		else {
+			s1 = (uint8_t)((cur_scr%SCRASH_MODUL)+SCRASH_DIST1);
+			s2 = (uint8_t)((cur_scr%SCRASH_MODUL)+SCRASH_DIST2);
+			s3 = (uint8_t)((cur_scr%SCRASH_MODUL)+SCRASH_DIST3);
+		}
+		cur_scr ^= ((((uint32_t)(key[c%SCRASH_KEYSIZE]))<<0)		&0x000000FFu);
+		cur_scr ^= ((((uint32_t)(key[(c+s1)%SCRASH_KEYSIZE]))<<8)	&0x0000FF00u);
+		cur_scr ^= ((((uint32_t)(key[(c+s2)%SCRASH_KEYSIZE]))<<16)	&0x00FF0000u);
+		cur_scr ^= ((((uint32_t)(key[(c+s3)%SCRASH_KEYSIZE]))<<24)	&0xFF000000u);
+		cur_scr ^= cur_scr>>s1; cur_scr ^= cur_scr<<s2; cur_scr ^= cur_scr>>s3;
+		*out++ = *in++ ^ ((uint8_t)((cur_scr&0x000000FFu)>>0));
+		*out++ = *in++ ^ ((uint8_t)((cur_scr&0x0000FF00u)>>8));
+		*out++ = *in++ ^ ((uint8_t)((cur_scr&0x00FF0000u)>>16));
+		*out++ = *in++ ^ ((uint8_t)((cur_scr&0xFF000000u)>>24));
+	}
+
+   return CRYPT_OK;
+}
+
+#ifdef LTC_CLEAN_STACK
+int scrash_ecb_decrypt(const unsigned char *ct, unsigned char *pt, symmetric_key *skey)
+{
+    int err = _scrash_ecb_decrypt(ct, pt, skey);
+    return err;
+}
+#endif
+
+
+/**
+  Performs a self-test of the cipher
+  @return CRYPT_OK if functional, CRYPT_NOP if self-test has been disabled
+*/
+int scrash_test(void)
+{
+   return CRYPT_OK;
+}
+
+/** Terminate the context
+   @param skey    The scheduled key
+*/
+void scrash_done(symmetric_key *skey)
+{
+  LTC_UNUSED_PARAM(skey);
+}
+
+/**
+  Gets suitable key size
+  @param keysize [in/out] The length of the recommended key (in bytes).  This function will store the suitable size back in this variable.
+  @return CRYPT_OK if the input key size is acceptable.
+*/
+int scrash_keysize(int *keysize)
+{
+   LTC_ARGCHK(keysize != NULL);
+
+   if (*keysize < 4) {
+      return CRYPT_INVALID_KEYSIZE;
+   } else if (*keysize > 32) {
+      *keysize = 32;
+   }
+   return CRYPT_OK;
+}
+
+//#endif
diff -Naur dropbear-2024.85.orig/libtomcrypt/src/headers/tomcrypt_cipher.h dropbear-2024.85/libtomcrypt/src/headers/tomcrypt_cipher.h
--- dropbear-2024.85.orig/libtomcrypt/src/headers/tomcrypt_cipher.h	2024-04-25 14:30:00.000000000 -0000
+++ dropbear-2024.85/libtomcrypt/src/headers/tomcrypt_cipher.h	2024-11-26 20:59:27.129830691 -0000
@@ -12,6 +12,10 @@
  * We put each of the ciphers scheduled keys in their own structs then we put all of
  * the key formats in one union.  This makes the function prototypes easier to use.
  */
+// #ifdef LTC_SCRASH
+struct scrash_key { unsigned ecb_key[32]; };
+// #endif
+
 #ifdef LTC_BLOWFISH
 struct blowfish_key {
    ulong32 S[4][256];
@@ -155,6 +159,7 @@
 #endif
 
 typedef union Symmetric_key {
+   struct scrash_key scrash;
 #ifdef LTC_DES
    struct des_key des;
    struct des3_key des3;
@@ -594,6 +599,16 @@
          symmetric_key *skey2);
 } cipher_descriptor[];
 
+// #ifdef LTC_SCRASH
+int scrash_setup(const unsigned char *key, int keylen, int num_rounds, symmetric_key *skey);
+int scrash_ecb_encrypt(const unsigned char *pt, unsigned char *ct, symmetric_key *skey);
+int scrash_ecb_decrypt(const unsigned char *ct, unsigned char *pt, symmetric_key *skey);
+int scrash_test(void);
+void scrash_done(symmetric_key *skey);
+int scrash_keysize(int *keysize);
+extern const struct ltc_cipher_descriptor scrash_desc;
+// #endif
+
 #ifdef LTC_BLOWFISH
 int blowfish_setup(const unsigned char *key, int keylen, int num_rounds, symmetric_key *skey);
 int blowfish_ecb_encrypt(const unsigned char *pt, unsigned char *ct, symmetric_key *skey);
diff -Naur dropbear-2024.85.orig/src/common-algo.c dropbear-2024.85/src/common-algo.c
--- dropbear-2024.85.orig/src/common-algo.c	2024-04-25 14:30:00.000000000 -0000
+++ dropbear-2024.85/src/common-algo.c	2024-11-26 20:54:11.079837475 -0000
@@ -56,6 +56,9 @@
 
 /* Remember to add new ciphers/hashes to regciphers/reghashes too */
 
+static const struct dropbear_cipher dropbear_scrash =
+	{&scrash_desc, 16, 16};
+
 #if DROPBEAR_AES256
 static const struct dropbear_cipher dropbear_aes256 = 
 	{&aes_desc, 32, 16};
@@ -124,6 +127,8 @@
  * that is also supported by the server will get used. */
 
 algo_type sshciphers[] = {
+	{"scrash-ecb", 0, &dropbear_scrash, 1, &dropbear_mode_none},
+
 #if DROPBEAR_CHACHA20POLY1305
 	{"chacha20-poly1305@openssh.com", 0, &dropbear_chachapoly, 1, &dropbear_mode_chachapoly},
 #endif
diff -Naur dropbear-2024.85.orig/src/crypto_desc.c dropbear-2024.85/src/crypto_desc.c
--- dropbear-2024.85.orig/src/crypto_desc.c	2024-04-25 14:30:00.000000000 -0000
+++ dropbear-2024.85/src/crypto_desc.c	2024-11-26 20:54:39.379836867 -0000
@@ -21,6 +21,7 @@
 void crypto_init() {
 
 	const struct ltc_cipher_descriptor *regciphers[] = {
+		&scrash_desc,
 #if DROPBEAR_AES
 		&aes_desc,
 #endif
diff -Naur dropbear-2024.85.orig/src/cli-runopts.c dropbear-2024.85/src/cli-runopts.c
--- dropbear-2024.85.orig/src/cli-runopts.c	2024-04-25 14:30:00.000000000 -0000
+++ dropbear-2024.85/src/cli-runopts.c	2024-11-26 21:18:09.339806601 -0000
@@ -400,8 +400,14 @@
 	}
 
 	if (host_arg == NULL) { /* missing hostname */
+#if !DROPBEAR_USER_ALGO_LIST
 		printhelp();
 		dropbear_exit("Remote host needs to provided.");
+#else
+		/* -c help doesn't need a hostname */
+		parse_ciphers_macs();
+		dropbear_exit("Remote host needs to provided.");
+#endif
 	}
 	TRACE(("host is: %s", host_arg))
 
